x-pack/plugins/observability/public/components/shared/exploratory_view/utils/telemetry.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { TrackEvent, METRIC_TYPE } from '../../../../hooks/use_track_metric';
+import type { SeriesUrl } from '../types';
+
+export const trackTelemetryOnApply = (
+  trackEvent: TrackEvent,
+  allSeries: SeriesUrl[],
+  reportType: string
+) => {
+  trackFilters(trackEvent, allSeries, reportType);
+  trackDataType(trackEvent, allSeries, reportType);
+  trackApplyChanges(trackEvent);
+};
+
+export const trackTelemetryOnLoad = (trackEvent: TrackEvent, start: number, end: number) => {
+  trackChartLoadingTime(trackEvent, start, end);
+};
+
+const getAppliedFilters = (allSeries: SeriesUrl[]) => {
+  const filtersByDataType = new Map<string, string[]>();
+  allSeries.forEach((series) => {
+    const seriesFilters = filtersByDataType.get(series.dataType);
+    const filterFields = (series.filters || []).map((filter) => filter.field);
+
+    if (seriesFilters) {
+      seriesFilters.push(...filterFields);
+    } else {
+      filtersByDataType.set(series.dataType, [...filterFields]);
+    }
+  });
+  return filtersByDataType;
+};
+
+const trackFilters = (trackEvent: TrackEvent, allSeries: SeriesUrl[], reportType: string) => {
+  const filtersByDataType = getAppliedFilters(allSeries);
+  [...filtersByDataType.keys()].forEach((dataType) => {
+    const filtersForDataType = filtersByDataType.get(dataType);
+
+    (filtersForDataType || []).forEach((filter) => {
+      trackEvent({
+        app: 'observability-overview',
+        metricType: METRIC_TYPE.COUNT,
+        metric: `exploratory_view__filters__filter_${filter}`,
+      });
+      trackEvent({
+        app: 'observability-overview',
+        metricType: METRIC_TYPE.COUNT,
+        metric: `exploratory_view__filters__report_type_${reportType}__data_type_${dataType}__filter_${filter}`,
+      });
+    });
+  });
+};
+
+const trackApplyChanges = (trackEvent: TrackEvent) => {
+  trackEvent({
+    app: 'observability-overview',
+    metricType: METRIC_TYPE.COUNT,
+    metric: 'exploratory_view_apply_changes',
+  });
+};
+
+const trackDataType = (trackEvent: TrackEvent, allSeries: SeriesUrl[], reportType: string) => {
+  const metrics = allSeries.map((series) => ({
+    dataType: series.dataType,
+    metricType: series.selectedMetricField,
+  }));
+
+  metrics.forEach(({ dataType, metricType }) => {
+    trackEvent({
+      app: 'observability-overview',
+      metricType: METRIC_TYPE.COUNT,
+      metric: `exploratory_view__report_type_${reportType}__data_type_${dataType}__metric_type_${metricType}`,
+    });
+  });
+};
+
+export const trackChartLoadingTime = (trackEvent: TrackEvent, start: number, end: number) => {
+  const secondsLoading = (end - start) / 1000;
+  switch (true) {
```

Comment:
This is just my opinion, but I think we'd reduce complexity by having a function whose only job is to create the appropriate `string`, and then having one line where we call `trackChartLoadingMetric`, rather than utilizing this clever `switch` block.

```typescript
function toRangeStr(n: number) {
  if (n < 0 || isNaN(n)) return null; // or `throw Error()`?
  if (n >= 60) return '60+';
  else if (n >= 30) return '30-60';
  else if (n >= 20) return '20-30';
  else if (n >= 10) return '10-20';
  else if (n >= 5) return '5-10';
  return '0-5';
}

// ...
trackChartLoadingMetric(trackEvent, toRangeStr(secondsLoading));
```